Source/NameServer/AddressBook/AddressBook.hpp

```diff
@@ -0,0 +1,165 @@
+
+    std::string TaskExecPath(std::string &TaskName);                     // Get the path of a task
+    unsigned TaskExecPath(std::string &TaskName, std::string &NewPath);  // Set the path of a task
+
+    bool TaskValid(std::string &TaskName);
+    bool TaskMapValid(std::string &TaskName);
+    bool TaskLinkValid(std::string &TaskName);
+
+    unsigned RebuildTask(std::string &TaskName);
+    unsigned BuildMaps(std::string &TaskName);
+    unsigned BuildLink(std::string &TaskName);
+
+    // Device Manipulation
+    unsigned AddDevice(std::string &TaskName, Record_t &Device, bool Validate = true);
+    unsigned UpdateDevice(std::string &TaskName, DeviceData_t &Data);
+
+    // Find a device
+    unsigned FindDevice(std::string &TaskName, std::string &Name, const Record_t* &DRec);
+    unsigned FindDevice(std::string &TaskName, SymAddr_t Address, const Record_t* &DRec);
+
+    // Find all devices supervised by X
+    unsigned FindBySuper(std::string &TaskName, SymAddr_t Supervisor, const RecordVect_t* &Records);
+
+    unsigned FindByType(std::string &TaskName, std::string Type, const RecordVect_t* &Records);
+    unsigned FindByAttribute(std::string &TaskName, std::string Attribute, const RecordVect_t* &Records);
+
+    unsigned FindByInMsg(std::string &TaskName, std::string Msg, const RecordVect_t* &Records);            //TODO
+    unsigned FindByOuMsg(std::string &TaskName, std::string Msg, const RecordVect_t* &Records);            //TODO
+
+    // device getters. Note that a vector<Record_t> is different from a RecordVect_t,
+    // which is a vector<const Record_t*>
+    unsigned GetDevices(std::string &TaskName, const std::vector<Record_t>* &Records);
+    unsigned GetExternals(std::string &TaskName, const std::vector<Record_t>* &Records);
+    unsigned GetSupervisors(std::string &TaskName, const std::vector<Record_t>* &Records);
+    unsigned GetExtCon(std::string &TaskName, const RecordVect_t* &Records);
+
+    // Device Count getters
+    long GetDeviceCount(std::string &TaskName);
+    long GetLoadedDeviceCount(std::string &TaskName);
+    long GetExternalCount(std::string &TaskName);
+    long GetLoadedExternalCount(std::string &TaskName);
+    int GetSupervisorCount(std::string &TaskName);
+    int GetMappedSupervisorCount(std::string &TaskName);
+
+    // Integrity
+    unsigned IntegCheck(void);                  // Integrity check of ALL tasks
+    unsigned IntegCheck(std::vector<unsigned> &TaskIntegs);
+    unsigned IntegTask(std::string &TaskName, bool Verbose = false, FILE * = stdout);  // Ingerity Check a Task by name
+  //unsigned IntegTask(const TaskRecord_t *TRec, bool Verbose = false, FILE * = stdout);     // Ingerity Check a Task by pointer
+
+    // Debugging
+    void Dump(std::string &Name);
+    void Dump(FILE * = stdout, std::string Name = "");
+
+private:
+    TaskMap_t TaskMap;
+    std::string ABderived; // changed to allow multiple inheritance (from CommonBase) 14 July 2019 ADR
```

Comment:
Advisory Board Derived? Also not sure what this is for (though it's read in `AddressBook::Dump`).

Edit: Ah, "address book". Still falling for that one.

Reply:
Changed to `AddressBookDerived`